### ISSUE TITLE
main.py error because no outputs folder

### ISSUE THREAD

**Issue by Michaeljurado24** [ezCGP/ezCGP#2]
Traceback (most recent call last):
  File "main.py", line 21, in <module>
    converged_solution = universe.create_universe(input_data=train_data, labels=train_labels, universe_seed=seed+i)
  File "D:\workspace\ezCGP\universe.py", line 119, in create_universe
    plt.savefig('outputs/seed%i_gen%i.png' % (universe_seed, generation))
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\pyplot.py", line 697, in savefig
    res = fig.savefig(*args, **kwargs)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\figure.py", line 1572, in savefig
    self.canvas.print_figure(*args, **kwargs)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\backend_bases.py", line 2244, in print_figure
    **kwargs)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\backends\backend_agg.py", line 550, in print_png
    filename_or_obj = open(filename_or_obj, 'wb')
FileNotFoundError: [Errno 2] No such file or directory: 'outputs/seed1_gen0.png'

You can fix this issue by adding an outputs folder or adding 
this line of code right before line 119 in universe.py: 
            from os import makedirs
            try:
                makedirs("outputs")
            except OSError as E:
                pass
            plt.savefig('outputs/seed%i_gen%i.png' % (universe_seed, generation))
Not sure if this is just a windows thing or what

**Comment by roddtalebi** [comment]
I'm going to try **[this](https://medium.com/@ageitgey/python-3-quick-tip-the-easy-way-to-deal-with-file-paths-on-windows-mac-and-linux-11a072b58d5f)** out so that relative filepaths work for all OS

tldr; use [pathlib](https://docs.python.org/3/library/pathlib.html)

**Comment by roddtalebi** [comment]
I just pushed some changes up; let me know if that works

**Comment by Michaeljurado24** [comment]
It broke on Windows and Ubuntu. I'm going to take a look at it:
 Traceback (most recent call last):
  File "main.py", line 21, in <module>
    converged_solution = universe.create_universe(input_data=train_data, labels=train_labels, universe_seed=seed+i)
  File "D:\workspace\ezCGP\universe.py", line 120, in create_universe
    Path('outputs').mkdir(parents=True, exists=True) #should help work on all OS
TypeError: mkdir() got an unexpected keyword argument 'exists'

**Comment by roddtalebi** [comment]
Should be 'exist_ok', not 'exists'

https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir

**Comment by Michaeljurado24** [comment]
After I changed that there was still an error with how the Path variable was used. This works on Linux and Windows:

import os
.......
os.makedirs(os.path.dirname("outputs/"), exist_ok=True)
filename = 'outputs/seed%i_gen%i.png' % (universe_seed, generation)
plt.savefig(filename)
plt.close()

**Comment by Michaeljurado24** [comment]
This appears to work as well if you Want to use Path
            Path('outputs').mkdir(parents=True, exist_ok =True) #should help work on all OS
            filename = 'outputs/seed%i_gen%i.png' % (universe_seed, generation)
            plt.savefig(filename)
            plt.close()

**Comment by roddtalebi** [comment]
What was the second error message you got though?

**Comment by Michaeljurado24** [comment]
Traceback (most recent call last):
  File "main.py", line 21, in <module>
    converged_solution = universe.create_universe(input_data=train_data, labels=train_labels, universe_seed=seed+i)
  File "D:\workspace\ezCGP\universe.py", line 122, in create_universe
    plt.savefig(filepath)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\pyplot.py", line 697, in savefig
    res = fig.savefig(*args, **kwargs)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\figure.py", line 1572, in savefig
    self.canvas.print_figure(*args, **kwargs)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\backend_bases.py", line 2244, in print_figure
    **kwargs)
  File "D:\Programs\Miniconda3\envs\tensorflow\lib\site-packages\matplotlib\backends\backend_agg.py", line 556, in print_png
    _png.write_png(renderer._renderer, filename_or_obj, self.figure.dpi)
TypeError: Object does not appear to be a 8-bit string path or a Python file-like object

The error stems from using Path('outputs/seed%i_gen%i.png' % (universe_seed, generation)) instead of a python file object or just a string.

**Comment by roddtalebi** [comment]
Bummer, looks like matplotlib doesn't accept Path...but then again maybe it will on v2.2?... I'll have to play around with it at home.

Seems like it could be good practice to use pathlib.Path or it's just over engineering...shrug emoji

https://github.com/matplotlib/matplotlib/issues/5968
https://github.com/matplotlib/matplotlib/issues/5968
